lib/install/management/configure.go

```diff
@@ -205,54 +311,50 @@ func (d *Dispatcher) Rollback(vch *vm.VirtualMachine, conf *config.VirtualContai
 		return errors.Errorf("could not complete manual rollback: %s", err)
 	}
 
-	return d.retryDeleteSnapshotByRef(&snapshot.Snapshot, snapshot.Name, conf.Name)
+	return d.deleteSnapshot(&snapshot.Snapshot, snapshot.Name, conf.Name)
 }
 
-// retryDeleteSnapshotByRef will retry to delete snapshot by its reference if there is GenericVmConfigFault returned. This is a workaround for vSAN delete snapshot
-func (d *Dispatcher) retryDeleteSnapshotByRef(snapshot *types.ManagedObjectReference, snapshotName, applianceName string) error {
-	// delete snapshot immediately after snapshot rollback usually fail in vSAN, so have to retry several times
-	operation := func() error {
-		return d.deleteSnapshotByRef(snapshot, snapshotName, applianceName)
-	}
-	var err error
-	if err = retry.Do(operation, isSystemError); err != nil {
-		d.op.Errorf("Failed to clean up appliance upgrade snapshot %q: %s.", snapshotName, err)
-		d.op.Errorf("Snapshot %q of appliance virtual machine %q MUST be removed manually before upgrade again", snapshotName, applianceName)
-	}
-	return err
-}
+// deleteSnapshot deletes the provided snapshot.  It retries on SystemError (GenericVmConfigFault) seen in vSAN
+func (d *Dispatcher) deleteSnapshot(snapshot *types.ManagedObjectReference, snapshotName, applianceName string) error {
+	defer trace.End(trace.Begin(fmt.Sprintf("Deleteing snaphost(%s) for %s", snapshotName, applianceName), d.op))
 
-func isSystemError(err error) bool {
-	if soap.IsSoapFault(err) {
-		if _, ok := soap.ToSoapFault(err).VimFault().(*types.SystemError); ok {
-			return true
+	// action function that will delete the snapshot in a retry loop
+	deleteSnap := func() error {
+		d.op.Infof("Deleting %s snapshot %q", d.Action.String(), snapshotName)
+		if _, err := d.appliance.WaitForResult(d.op, func(ctx context.Context) (tasks.Task, error) {
+			consolidate := true
+			return d.appliance.RemoveSnapshot(d.op, snapshot, false, &consolidate)
+		}); err != nil {
+			return err
 		}
+		return nil
 	}
 
-	if soap.IsVimFault(err) {
-		if _, ok := soap.ToVimFault(err).(*types.SystemError); ok {
-			return true
+	// retry on any general SystemError
+	isSystemError := func(err error) bool {
```

Comment:
should we retry on the transient "TaskInProgress" or similar errors as well?

Reply:
Possibly...I thought about just using the `tasks.WaitForResult` and adding `SystemError` -- this was older code so _assumed_ that if that was needed it _might_ have surfaced in testing or user experience...happy to change if you think it's worthwhile...

I just glanced at the API documentation around the task...and I think it would be best to use `tasks.WaitForResult` as it should better protect us...

Reply:
FYI - @AngieCris has added an isTransientError or similar as part of the API list work that just merged. If you chose to update this you may find that useful.